lib/fdotypes.c

```diff
@@ -5471,6 +5471,110 @@ bool fdo_serviceinfo_modules_list_write(fdow_t *fdow) {
 	return ret;
 }
 
+/**
+ * Return bool value representing whether any external ServiceInfo module has message to
+ * send in the NEXT iteration. This determines the TO2.DeviceServiceInfo.IsMoreServiceInfo value
+ * for any currently active module.
+ *
+ * @param fdow - Pointer to the writer.
+ * @param module_list - Pointer to the ServiceInfo module list containing all
+ * Device ServiceInfos modules.
+ * @param mtu - MTU to be used for fitting the values.
+ *
+ * @return true if there is ServiceInfo to be sent in the next iteration,
+ * false on any other conditions.
+ */
+bool fdo_serviceinfo_external_mod_is_more(fdow_t *fdow,
+	fdo_sdk_service_info_module_list_t *module_list, size_t mtu) {
+
+	if (!fdow || !module_list) {
+		return false;
+	}
+	fdo_sdk_service_info_module_list_t *traverse_list = module_list;
+	bool is_more = false;
+
+	while (traverse_list) {
+		if (traverse_list->module.active &&
+			traverse_list->module.service_info_callback(
+			FDO_SI_IS_MORE_DSI, NULL, fdow, NULL, NULL, &is_more, mtu) != FDO_SI_SUCCESS) {
+			LOG(LOG_DEBUG, "Sv_info: %s's CB Failed for type:%d\n",
+			    traverse_list->module.module_name, FDO_SI_HAS_MORE_DSI);
+			return false;
+		}
+		if (is_more) {
+			return is_more;
+		}
+		traverse_list = traverse_list->next;
+	}
+	return false;
+}
+
+/**
+ * Return a module reference that has some ServiceInfo to be send NOW/immediately,
+ * by making a callbacks to each active module, to determine whether the module
+ * has something to send immediately.
+ *
+ * @param fdow - Pointer to the writer.
+ * @param module_list - Pointer to the ServiceInfo module list containing all
+ * Device ServiceInfos modules.
+ * @param mtu - MTU to be used for fitting the values.
+ *
+ * @return Pointer to/module reference (fdo_sdk_service_info_module *) if there is any module
+ * that has ServiceInfo to send NOW/immediately, else return NULL.
+ */
+fdo_sdk_service_info_module* fdo_serviceinfo_get_external_mod_to_write(fdow_t *fdow,
+	fdo_sdk_service_info_module_list_t *module_list, size_t mtu) {
+
+	if (!fdow || !module_list) {
+		return false;
+	}
+	fdo_sdk_service_info_module_list_t *traverse_list = module_list;
+	bool has_more = false;
+
+	while (traverse_list) {
+		if (traverse_list->module.active &&
+			traverse_list->module.service_info_callback(
+			FDO_SI_HAS_MORE_DSI, NULL, fdow, NULL, &has_more, NULL, mtu) != FDO_SI_SUCCESS) {
+			LOG(LOG_DEBUG, "Sv_info: %s's CB Failed for type:%d\n",
+			    traverse_list->module.module_name, FDO_SI_HAS_MORE_DSI);
+			return false;
```

Comment:
Should this be handled differently that a is_more == false condition? In other words, do we need to do something more than log an error?

Reply:
This should've always returned NULL. Updated.